### PR TITLE
[WIP] ColorBar visual

### LINKED ISSUE #682: Colorbar visual
* with/without border
* ticks
* tick labels

colorbar

### PR DESCRIPTION
**Pull request by bollu**
closes #682 

Started working on this, learnt a couple of things about vispy's model in the process. I've got the basic structure down.

- [x] add a ```orientation``` field for ```horizontal``` or ```vertical```
- [x] add a ```label``` field
- [x] ~~add a ```ticks``` field for the colorbar~~
- [ ] add an accompanying node for the SceneGraph
- [ ] draw a customizable border around the colorbar
- [ ] add test cases for the colorbar
- [x] actually respect the ```clim``` parameter (this should be easy, just change the ```texcoords```
- [x] document the example, so that there's proper explanation
- [x] write documentation (docstrings) for the files
- [x] add a license notice for the files

***Screenie of colorbar :***

screenshot from 2015-06-04 23 48 12

### CODE REVIEW

**Comment by Eric89GXL** [comment]
I don't think you need to do anything special to respect the `clim` bounds -- `clim` sets what the start and end of the colorbar mean for the data, so you should be able to just plot the bar, and then label the start and end of it with `clim[0]` and `clim[1]` for now.

**Comment by Eric89GXL** [comment]
And let's not bother with ticks, we should eventually do that using the `AxisVisual` somehow

**Comment by bollu** [comment]
@Eric89GXL: okay, I'm confused, doesn't clim give you the start and end range for the colorbar interpolation? like, by default, you go from ```colormap([0, 1])```, but with clim, you go from ```colormap([clim_start, clim_end])``` correct?

**Comment by Eric89GXL** [comment]
No, think about the common use case. For example, I have some data that ranges from -100 to 1000. But when I plot it I clip it a little bit, so I do:
```Python
>>> clim = [-5, 900]
>>> cmap = 'cubehelix'
>>> ax.image(..., colormap=cmap, clim=clim)
```
It would be nice to be able to then do:
```
ax.colorbar(cmap, clim)
```
and have it "do the right thing", which in this case is to display the full colorbar, with the left side marked as -5 and the right side marked as 900. Does that make sense?

**Comment by Eric89GXL** [comment]
In other words, I don't think the most common use case will be to "cut into" part of the colormap for display, but rather to use the entire colormap, but have the start and end points mean something different for the data.

I suppose eventually we'll also have to figure out what happens if they do want to "cut into" the colormap, or only display a subsection of the colormap, or if the colormap is scaled in some nonlinear way... but we should be able to tackle that later once visuals actually support that sort of thing.

**Comment by bollu** [comment]
So, in that case, the range ```[-100, 1000]``` gets linearly mapped to the range ```[-5, 100]``` ? or is the data clamped? ```[-100, -5] -> { -5 }```, ```[-5, 900]``` remains identical, ```[900, 1000] -> { 900 }```

(I think clamping is the right behaviour?)

So, in case of clamping, even the colors get clamped? like, ```[-100, -5] -> 0```, ```[-5, 900] -> [0, 1]```, ```[900, 1000] -> 1```?

**Comment by Eric89GXL** [comment]
Yeah, colors outside the range get clamped.

**Comment by bollu** [comment]
@Eric89GXL so, there's nothing "extra" to be done by the colorbar, right? like, it doesn't need to modify the color rendering, just rendering linearly would work?

**Comment by Eric89GXL** [comment]
Yep, should be pretty basic for now

**Comment by bollu** [comment]
@Eric89GXL how would I take the label for the colorbar? as a string? if so, how does one allow styling of the label?

**Comment by Eric89GXL** [comment]
the label should be a string, yeah. We should probably eventually allow styling during construction, but this might blow up the constructor arguments which is not good. It's probably easier to expose the underlying `label` as an attribute, where `colorbar.label` is an instance of `TextVisual`, then people can change properties as they wish

**Comment by bollu** [comment]
@Eric89GXL ah, yes. That does sound like a nice way to go about it. I'll do it that way.

**Comment by bollu** [comment]
@Eric89GXL: I can't really use the ```clim``` field unless there's an axis to it, right? So do I simply leave it as a dummy field?

**Comment by Eric89GXL** [comment]
For now just place labels at the start and end using a single TextVisual object, eventually we will replace it with an `AxisVisual` but we can do that later

**Comment by bollu** [comment]
is there anyway to get the bounding box / size of a ```TextVisual```? I don't really see a straightforward way of doing this. 

When I place the label, I need to make sure that it doesn't intersect with the ```ColorBar``` whatever the dimensions of the ```ColorBar``` are, or whatever font size the ```TextVisual``` uses. I'd like to query the ``TextVisual``` for it's dimensions for this reason.

If this is not possible, what is a reasonable compromise when it comes to placing the labels?

**Comment by QuLogic** [comment]
@bollu I think what you want is to set the `TextVisual`'s `anchor_x` and `anchor_y` settings to something other than `'center'` (though that may be a bit broken #975).

**Comment by bollu** [comment]
@QuLogic: thank you. That worked.

**Comment by bollu** [comment]
@Eric89GXL should the border around the ColorBar measured in pixels? I'm assuming it should be.

**Comment by Eric89GXL** [comment]
Yep

**Comment by bollu** [comment]
@Eric89GXL that will require doing the calculation in a coordinate system measured in pixels. I'm assuming that's the ```Visual``` coordinate system?

**Comment by Eric89GXL** [comment]
You want the `document` coordinate system. Look through this example:

https://github.com/vispy/vispy/blob/master/examples/tutorial/visuals/T02_measurements.py

I strongly recommend going through each of those examples, they will save you some pain when developing visuals.

**Comment by bollu** [comment]
current screenshot:

screenshot from 2015-06-07 23 51 30

It's not very flattering, but it works :)

File: vispy/visuals/colorbar.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            # https://github.com/vispy/vispy/issues/981
+            self._ticks = []
+            self._ticks.append(TextVisual(str(clim[0]),
+                               pos=begin_tick_pos,
+                               anchor_y="top"))
+            self._ticks.append(TextVisual(str(clim[1]),
+                               pos=end_tick_pos,
+                               anchor_y="top"))
+
+        elif orientation == "vertical":
```
Comment (bollu) [inline vispy/visuals/colorbar.py]: I recognize that there is code duplication here, but I prefer the explicitness of it, since the two halves of the code setup the colorbar differently. (```rotation``` is setup, along with different positions for the markers).

I'd like your opinions.

**Comment by bollu** [comment]
@Eric89GXL the test cases for visuals do a direct pixel-by-pixel comparison? ```assert_image_approved``` keeps popping up in the test cases. So, if I'm going to write tests, I'll need to use the same function to do it?

**Comment by Eric89GXL** [comment]
Yeah, but I have a feeling the colorbar might change a little bit when using the AxisVisual, so for now just add a `XXX TODO`, and just test that it actually works and renders something non-zero.

**Comment by Eric89GXL** [comment]
But yes, eventually you'll want to use the image comparison functions. There is a cross-correlation type mode, and a pixel-difference type mode. Either should eventually work for this visual.

File: vispy/visuals/colorbar.py (hunk 1 above)
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: I think it's okay to have this level of redundancy. But I'd change it to `else:  # orientation == 'vertical':`, that way 1) we save one comparison, 2) it's clear that we know that to be the state at this point, and 3) it's easy to add the next `elif` mode by deleting a few chars.

**Comment by bollu** [comment]
@Eric89GXL cool, so, add a TODO and check if it renders _anything_?

**Comment by Eric89GXL** [comment]
Yeah for now that should be okay. Not ideal but it will remind us to check it for real later. If you really want, you can add the full test now, and modify it later.

**Comment by bollu** [comment]
New screenshot (picked flattering colors :) )

screenshot from 2015-06-09 06 29 09

File: vispy/visuals/colorbar.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        tex_coords = np.array([[0, 0], [1, 0], [1, 1],
+                              [0, 0], [1, 1], [0, 1]],
+                              dtype=np.float32)
+
+        self._program.frag['color_transform'] = Function(self._cmap.glsl_map)
+        self._program['a_texcoord'] = tex_coords.astype(np.float32)
+
+        self.regenerate()
+
+    def regenerate(self):
```
Comment (bollu) [inline vispy/visuals/colorbar.py]: I'm not a fan of this name, but I couldn't think of anything better than ```regenerate```. Do you have any suggestions?

Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: when does it need to be called? it should probably be private for now

**Comment by Eric89GXL** [comment]
Nice example! Can you e.g. show an Image with a colormap, and put colorbars on each side of it? That would make a potential real-life use case clearer :)

File: vispy/visuals/colorbar.py (hunk 2 above)
Comment (bollu) [inline vispy/visuals/colorbar.py]: Ah, yes. It should have been private in the first place.

Well, whenever you need to recreate shader data or recreate the label or
ticks or something, it's called.

So, for example, when you change the border width, the positions of the
labels need to be regenerated to not overlap with the new border.

**Comment by bollu** [comment]
Sure :) that was mostly me messing around with it and changing a bunch of
properties to get something pretty.

File: vispy/visuals/colorbar.py (hunk 2 above)
Comment (QuLogic) [inline vispy/visuals/colorbar.py]: So, simple old `_update` then?

Comment (bollu) [inline vispy/visuals/colorbar.py]: @QuLogic that works too :) But, I felt that ```_update``` was too generic for some reason.

Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: as long as it doesn't collide with any super-class properties, sure -- but be sure to check if you use that name

File: vispy/visuals/colorbar.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        self._border_program['a_position'] = vertices.astype(np.float32)
+
+        self._border_program.vert['border_width'] = self._border_width
+        self._border_program.frag['border_color'] = self._border_color.rgba
+
+    @staticmethod
+    def _get_orientation_error(orientation):
+        return ValueError("orientation must"
+                          " be \'horizontal\' or \'vertical\'."
+                          " given: %s" % orientation)
```
Comment (QuLogic) [inline vispy/visuals/colorbar.py]: I would say `... or 'vertical', not '%s'."` Also, the single quotes do not need to be escaped. Finally, it is safer to encapsulate `orientation` in a tuple as `"..." % (orientation, )`.

Comment (bollu) [inline vispy/visuals/colorbar.py]: noted. Will change!

**Comment by bollu** [comment]
@Eric89GXL are there any canonical colormap images I could use?

**Comment by Eric89GXL** [comment]
Not sure. I'd probably just construct something artificial like:
```
y, x = np.meshgrid(np.linspace(-1, 1, 30), np.linspace(0, 2, 20))
data = np.exp(-(x*x + y*y))
```
And then do a colormap'ed image of `data`

**Comment by bollu** [comment]
Decided to plot the mandlebrot set for more authenticity :)

screenshot from 2015-06-09 19 42 48


@Eric89GXL is this okay as an example?

**Comment by rougier** [comment]
Nice !

**Comment by Eric89GXL** [comment]
Yes, LGTM :)

Maybe we should have orientation = 'left' | 'right' | 'top' | 'bottom', which would set the tick directions accordingly to always be toward the outside -- it seems like the label and ticks in your example should be pointing the other direction to me

File: vispy/visuals/colorbar.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+            self._label.pos = text_x, text_y
+
+            # TODO, HACK: This should ideally be a single TextVisual
+            # However, one TextVisual with multiple strings
+            # does not seem to be working as of now. (See #981)
+            # https://github.com/vispy/vispy/issues/981
+            self._ticks[0].pos = x - halfw, text_y
+            self._ticks[1].pos = x + halfw, text_y
+
+        elif self._orientation == "top":
```
Comment (bollu) [inline vispy/visuals/colorbar.py]: Number of branches has gone from 2 to 4. This piece of code needs some love, but I'm not sure as to _how_ to rewrite it.

File: vispy/visuals/colorbar.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+        self._clim = clim
+        # new TextVisuals need to be created
+        self._update()
+
+    @property
+    def label(self):
+        return self._label
+
+    @label.setter
+    def label(self, label):
```
Comment (bollu) [inline vispy/visuals/colorbar.py]: As of now, I'm letting the user have direct access to the ```label``` ```TextVisual``` that's present inside the colorbar. This breaks encapsulation. However, if we want to allow styling, it'll require more getters / setters (```set_label_color```, ```set_label_font_size```,...) What's the right compromise?

File: vispy/visuals/colorbar.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+        self._label = label
+        # position, styling has to be applied
+        self._update()
+
+    @property
+    def ticks(self):
```
Comment (bollu) [inline vispy/visuals/colorbar.py]: Same problem as that of ```label```. I'm giving the user direct access, which I'm not too fond of

File: examples/basics/visuals/colorbar_visual_types.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+colormap = get_colormap("ice")
+
+
+def style_colorbar(colorbar):
+    colorbar.label.font_size = 18
+    colorbar.label.color = "black"
+
+    colorbar.clim = (0, 10)
+
+    colorbar.ticks[0].font_size = 14
```
Comment (bollu) [inline examples/basics/visuals/colorbar_visual_types.py]: setting ticks needs you to access them individually. This should change once ```AxisVisual``` stabilizes, but until then, is there anyway to make this nicer? (overload the ```=``` operator by returning a custom ```Ticks``` class came to my mind, but it felt like over-engineering)

**Comment by bollu** [comment]
@Eric89GXL  Do the new orientations look correct to you?
screenshot from 2015-06-10 16 27 51

Also, I reworked the example with the new orientation:

screenshot from 2015-06-10 16 28 48


I think this is stable enough to write a testcase against. What do you think?

**Comment by rossant** [comment]
in the last example i'd rather see the text going from bottom to top

**Comment by bollu** [comment]
@rossant hm, that's fair. Will change

**Comment by bollu** [comment]
@rossant  pushed the change. Does this work?
screenshot from 2015-06-10 18 15 22
screenshot from 2015-06-10 18 15 04

**Comment by rossant** [comment]
it seems more natural to me like this, but let's wait for others' opinions!

**Comment by bollu** [comment]
@rossant, yeah, you're right. If I was drawing a graph on paper as well, that would be way I would label the y axis. It does feel more natural.

@Eric89GXL thoughts?

**Comment by bollu** [comment]
Also, @Eric89GXL  shall I start writing test-cases for this?

File: vispy/visuals/colorbar.py (hunk 4 above)
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: usually there is some way to set up parameters (scale factors, variable names) in the branches, and then set all the attributes after the branching. see if that's possible here

File: vispy/visuals/colorbar.py (hunk 5 above)
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: don't bother. eventually we can just give them axis to the ticks somehow, but it will probably change once we change it to use the axis visual

File: examples/basics/visuals/colorbar_visual_types.py (hunk 7 above)
Comment (Eric89GXL) [inline examples/basics/visuals/colorbar_visual_types.py]: I would say for now, don't allow setting these things to keep it simpler

**Comment by Eric89GXL** [comment]
you can write tests, but they will probably need to change quickly. it's probably better to focus instead of translating visuals, and then once that's done finishing axes, once that's done then making this use that visual, then once that's done it's probably stable enough to write some tests :)

**Comment by jdreaver** [comment]
@rossant @bollu +1 on the bottom-to-top text orientation :+1:

**Comment by Eric89GXL** [comment]
@bollu perhaps just add some basic cases that make sure it renders for the various orientations and choices of limits, and we can improve the tests later. that way we can get this merged and you can move on to other things.

**Comment by bollu** [comment]
please merge the [test data](https://github.com/vispy/test-data/pull/2) first so the CI passes

**Comment by bollu** [comment]
@Eric89GXL. @rossant: I feel that this is okay to get reviewed now.

File: examples/basics/visuals/colorbar_visual.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+# Copyright (c) 2015, Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+
+"""Using Colorbars with the Canvas"""
+
+from vispy import app
+from vispy import gloo
+from vispy.visuals.transforms import NullTransform, STTransform
+from vispy.visuals.transforms import TransformSystem
```
Comment (Eric89GXL) [inline examples/basics/visuals/colorbar_visual.py]: By convention we group imports like this if we need more lines
```
from vispy.visuals.transforms import (a, b,
                                      c)
```

File: vispy/testing/image_tester.py [hunk 9]
```diff
@@ -74,6 +74,9 @@ def assert_image_approved(image, standard_file, message=None, **kwargs):
     This function will automatically clone the test-data repository into 
     ~/.vispy/test-data. However, it is up to the user to ensure this repository
     is kept up to date and to commit/push new images after they are saved.
+
+    Run the test with python <test-path> --vispy-audit-tests to bring up
+    the auditing GUI.
```
Comment (Eric89GXL) [inline vispy/testing/image_tester.py]: thanks for improving the docs

File: vispy/visuals/colorbar.py [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+# visually "sticking" to the Colorbar
+_TEXT_PADDING_FACTOR = 1.2
+
+
+class ColorBarVisual(Visual):
+    """Visual subclass displaying a colorbar
+
+    Parameters
+    ----------
+
```
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: no empty newline here

File: vispy/visuals/colorbar.py [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+    center_pos : tuple (x, y)
+        Position where the colorbar is to be placed with
+        respect to the center of the colorbar
+    halfdim : tuple (half_width, half_height)
+        Half the dimensions of the colorbar measured
+        from the center. That way, the total dimensions
+        of the colorbar is (x - half_width) to (x + half_width)
+        and (y - half_height) to (y + half_height)
+    cmap : str | vispy.color.ColorMap
+        either the name of the ColorMap to be used from the standard
```
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: start with capital letter

File: vispy/visuals/colorbar.py [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+        set of names (refer to `vispy.color.get_colormap`),
+        or a custom ColorMap object.
+
```
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: no need for this extra empty newline

File: vispy/visuals/colorbar.py [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+        The label that is to be drawn with the colorbar
+        that provides information about the colorbar.
+    clim : tuple (min, max)
+        the minimum and maximum values of the data that
+        is given to the colorbar. This is used to draw the scale
+        on the side of the colorbar.
+    border_width : float (in px)
+        The width of the border the colormap should have. This measurement
+        is given in pixels
+    border_color : string | vispy.color.Color
```
Comment (Eric89GXL) [inline vispy/visuals/colorbar.py]: s/string/str/g

File: vispy/testing/image_tester.py (hunk 9 above)
Comment (bollu) [inline vispy/testing/image_tester.py]: you're welcome :)

**Comment by Eric89GXL** [comment]
@bollu you moving on to helping @campagnola with #928 by translating visuals next? That would probably be the highest impact move since a) you'll learn about a bunch of visuals that way and b) that PR should help improve performance quite a bit.

**Comment by campagnola** [comment]
FYI I have text, axis, and plotline visuals in progress.

**Comment by bollu** [comment]
@campagnola I was checking out the ```LineVisual``` example and was planning to port ```polygon``` as you suggested

**Comment by Eric89GXL** [comment]
Thanks for this @bollu, this should be useful with the plotting interface. Hopefully the old visuals system makes sense, and the new one won't be too hard to learn.

**Comment by bollu** [comment]
@Eric89GXL I'm reading through the new visuals code :) It seems pretty cool, especially with the shared shaders and everything.
